Modules/_testcapi/file.c

```diff
@@ -57,9 +57,50 @@ _testcapi_py_fopen_impl(PyObject *module, PyObject *path, const char *mode,
 }
 
 
+/*[clinic input]
+_testcapi.py_universalnewlinefgets
+
+    file: object
+    size: int
+    /
+
+Read a line from a file using Py_UniversalNewlineFgets.
+[clinic start generated code]*/
+
+static PyObject *
+_testcapi_py_universalnewlinefgets_impl(PyObject *module, PyObject *file,
+                                        int size)
+/*[clinic end generated code: output=2ce1bc76c9dc871c input=02c236049d18569a]*/
+{
+    int fd = PyObject_AsFileDescriptor(file);
+    if (fd == -1) {
+        return NULL;
+    }
+    FILE *fp = fdopen(fd, "rb");
+    if (fp == NULL) {
+        PyErr_SetFromErrno(PyExc_OSError);
+    }
+
+    char *buf = (char *)PyMem_Malloc(size);
+    if (buf == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    char *result = Py_UniversalNewlineFgets(buf, size, fp, NULL);
```

Comment:
fp can now be closed

Reply:
on `fclose(fp);` it gives me

----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../cpython/Lib/test/test_capi/test_file.py", line 305, in test_py_universalnewlinefgets
    with open(filename, "rb") as fp:
         ~~~~^^^^^^^^^^^^^^^^
OSError: [Errno 9] Bad file descriptor